### DISCUSSION TITLE
Two assertion mechanisms

### DISCUSSION THREAD

**Discussion by Vort** [bochs-emu/Bochs#157] [Q&A] [answered]
Please explain why Bochs code uses both `assert()` and `BX_ASSERT()` macros.
What is the logic behind choosing one or another option?

I'm asking because I was surprised to see `assert`s disabled in MSVC Debug builds.
But in case of turning them on, should `BX_ASSERT`s be turned on as well or not?

**Comment by stlintel** [comment]
I didn't know classic C++ assert can be disabled for speed.
BX_ASSERT was invented to be enabled in developer build but disabled in release s.t. it won't affect simulation speed.

**Comment by Vort** [comment]
> I didn't know classic C++ assert can be disabled for speed.

This is what `NDEBUG` define does here and in similar places: 
https://github.com/bochs-emu/Bochs/blob/5147c20eaa03f67f4fa87916c58b1fc58cd0ab18/bochs/build/win32/vs2019-workspace/vs2019/bochs.vcxproj#L140-L142
_(I disabled `NDEBUG` for this screenshot)_
![image](https://github.com/bochs-emu/Bochs/assets/1242858/0a22fcdc-6970-4dd4-aa37-71471e9620df)

> BX_ASSERT was invented to be enabled in developer build but disabled in release s.t. it won't affect simulation speed.

Can the same logic be applied to `assert`?
I think it is better to get rid of `NDEBUG` in Debug builds, this is how MSVC projects are usually configured.

**Comment by stlintel** [comment, marked as answer]
This NDEBUG is not present in gcc and I even don't use MSVCPP.
So I think you can get rid of NDEBUG if you want :)

I coded it with the state of mind like this:
Some asserts are really critical and must be done always and never disabled. These are classic assert().
Others affect performance too much or not interesting outside of new code development -> these are BX_ASSERT

**Comment by Vort** [comment]
> So I think you can get rid of NDEBUG if you want :)

Ok, I made PR with this change: #158
